content/md/posts/clojure-cli-tools-understanding-aliases.md

```diff
@@ -0,0 +1,222 @@
+
+In this example, a `"--middleware"` flag has multiple values which are wrapped in a Clojure vector (adding cider and refactor-nrepl middleware).  The `"-i"` flag has no value associated with it (it runs an interactive REPL prompt)
+
+```clojure
+  :repl/cider-refactor
+  {:extra-deps {nrepl/nrepl                   {:mvn/version "0.9.0"}
+                cider/cider-nrepl             {:mvn/version "0.27.4"}
+                refactor-nrepl/refactor-nrepl {:mvn/version "3.1.0"}}
+   :main-opts  ["-m" "nrepl.cmdline"
+                "--middleware" "[refactor-nrepl.middleware/wrap-refactor,cider.nrepl/cider-middleware]"
+                "-i"]}
+```
+
+> Use a `,` comma character to represent a space in an alias value.  The comma will be interpreted by Clojure as a space.
+
+## Alias definition for clojure.exec
+
+Aliases for clojure.exec are defined in deps.edn files
+
+`:exec-fn` specifies the fully qualified name of the function.
+
+`:exec-args` specifies a hash-map of default key/value pairs passed to the `:exec-fn` function.  The defaults can be overriden on the command line.
+
+Arguments can be nested within the `:exec-args` map, especially useful on projects with several component configurations (server, database, logging) and managed by a component system (i.e Integrant)
+
+```clojure
+{:aliases
+ {:project/run
+  {:exec-fn practicalli.service/start
+   :exec-args {:http/server {:port 8080
+                             :join? fale}
+               :log/mulog :elastic-search}}}}
+```
+
+Arguments in a nested map within the alias can be traversed (as with `get-in` and `update-in` functions) to override the default values in the alias.  So to set a different port value :
+
+```bash
+clojure -X:project/run '[:http/server :port]' 8888 :log/mulog :console :profile :dev
+```
+
+In this command the vector defines the path to the `:port` key and over-rides the default value. :log/mulog is a top-level key which changes the log publisher type.  `:profile` is another top-level key that sets the environment to `:dev` (e.g. to configure Integrant / Aero).
+
+Argument keys should either be a top-level key or a vector of keys to refer to a key in a nested hash-map of arguments.
+
+Key/value pairs passed on the command line override any of the argument values define in the `:exec-args` section of the respective alias.
+
+> An alias can contain configuration to run both `clojure.main` and `clojure.exec` (useful if steadily migrating users from -M to -X approach without breaking the user experience)
+
+
+### Form of clojure.exec command line arguments
+
+Key/value pairs are read as edn strings.
+
+The command line shell needs a little help parsing
+
+Some EDN arguments need to be wrapped in single quotes, `''` to prevent the command line shell incorrectly interpreting them.  Number values and keywords should not need wrapping though.
```

Comment:
That's pretty vague. Probably better to say that vectors and hash maps need to be wrapped in single quotes to avoid the shell trying to splitting arguments at spaces, e.g., `'[:a :b]'`, `'{:c 1}'`